Sqlite tutorial

DominicFrei opened the pull request.

LaPeste commented on tutorials/persistance-comparison-series/Sqlite/sqlite.md:
```diff
@@ -0,0 +1,390 @@
+
+Our journey of exploring options given to use when it comes persistence in Unity will in this part lead to databases. More specificaclly: Sqlite.
+
+Sqlite is a C based database that is used in many areas. It has been around for a long time and also found its way into the Unity world. During this tutorial series we have seen options like `PlayerPrefs` in Unity and on the other side `File` and `BinaryWriter`/`BinaryReader` provided by the unterlying .NET framework.
```
```suggestion
Sqlite is a C based database that is used in many areas. It has been around for a long time and also found its way into the Unity world. During this tutorial series we have seen options like `PlayerPrefs` in Unity and on the other side `File` and `BinaryWriter`/`BinaryReader` provided by the underlying .NET framework.
```

LaPeste commented on tutorials/persistance-comparison-series/Sqlite/sqlite.md:
```diff
@@ -0,0 +1,390 @@
+- Part 5: Realm Unity SDK *(coming soon)*
+- Part 6: Comparison of all those options
+
+Similar to the previous parts, this tutorial can also be found in our [Unity examples repository](the https://github.com/realm/unity-examples) on the [persistence-comparison](https://github.com/realm/unity-examples/tree/persistence-comparison) branch.
```
```suggestion
Similar to the previous parts, this tutorial can also be found in our [Unity examples repository](https://github.com/realm/unity-examples) on the [persistence-comparison](https://github.com/realm/unity-examples/tree/persistence-comparison) branch.
```

LaPeste commented on tutorials/persistance-comparison-series/Sqlite/sqlite.md:
```diff
@@ -0,0 +1,390 @@
+
+When opening it, the first thing we want to do, is importing Sqlite by adding `using Mono.Data.Sqlite;` and `using System.Data;` at the top of the file (1).
+
+Next we will look at how to save whenever the hit count is changed, which happens during `OnMouseDown()`. Similar to the `FileStream` and the `BinaryWriter` we have seen in the previous parts of this series we need some way to write the data to where it should go. In case of Sqlite that means we need a connection to the database. This is offered by the Sqlite library via the [`IDbConnection`](htt ...
```
```suggestion
Next we will look at how to save whenever the hit count is changed, which happens during `OnMouseDown()`. Similar to the `FileStream` and the `BinaryWriter` we have seen in the previous parts of this series we need some way to write the data to where it should go. In case of Sqlite that means we need a connection to the database. This is offered by the Sqlite library via the [`IDbConnection`](https://docs.microsoft.com/en-us/dotnet/api/system.data.idbconnection?view=net-6.0) class (2) which represents an open connection to the database. Since we will need a connection for loading the data later on again, we will extract opening a database connection into another function and call it `private IDbConnection CreateAndOpenDatabase()` (3).
```

LaPeste replied:
> Similar to the `FileStream` and the `BinaryWriter` we have seen in the previous parts of this series we need some way to write the data to where it should go. In case of Sqlite that means we need a connection to the database

I'd just remove this. And go like: In order to save date in Sqlite we first need to open a connection to the database.

Something along these lines.

LaPeste commented on tutorials/persistance-comparison-series/Sqlite/sqlite.md:
```diff
@@ -0,0 +1,390 @@
+
+Next we will look at how to save whenever the hit count is changed, which happens during `OnMouseDown()`. Similar to the `FileStream` and the `BinaryWriter` we have seen in the previous parts of this series we need some way to write the data to where it should go. In case of Sqlite that means we need a connection to the database. This is offered by the Sqlite library via the [`IDbConnection`](htt ...
+
+In there we first define a name for our database file. I'll just call it `MyDatabase` for now, accordingly the URI should be `"URI=file:MyDatabase.sqlite"` (4). Then we can create a connection this database using `new SqliteConnection(dbUri)` (5) and open it with `dbConnection.Open()` (6).
```
```suggestion
In there we first define a name for our database file. I'll just call it `MyDatabase` for now, accordingly the URI should be `"URI=file:MyDatabase.sqlite"` (4). Then we can create a connection to this database using `new SqliteConnection(dbUri)` (5) and open it with `dbConnection.Open()` (6).
```

LaPeste commented on tutorials/persistance-comparison-series/Sqlite/sqlite.md:
````diff
@@ -0,0 +1,390 @@
+}
+```
+
+Now we can work with this Sqlite database. Before we can actually add data to the it though, we need to set up a structure. This means, creating and defining tables which is the way most databases are organized. The following screenshot shows the final state we will create in this example.
````
```suggestion
Now we can work with this Sqlite database. Before we can actually add data to it though, we need to set up a structure. This means, creating and defining tables which is the way most databases are organized. The following screenshot shows the final state we will create in this example.
```

LaPeste commented on tutorials/persistance-comparison-series/Sqlite/sqlite.md:
```diff
@@ -0,0 +1,390 @@
+
+<img src="images/06_sqlite_tables.png" alt="Sqlite tables"/>
+
+When accessing or modifying the database we use [`IDbCommand`](https://docs.microsoft.com/en-us/dotnet/api/system.data.idbcommand?view=net-6.0) (6) which represents and SQL statement that can be executed on a database.
```
```suggestion
When accessing or modifying the database we use [`IDbCommand`](https://docs.microsoft.com/en-us/dotnet/api/system.data.idbcommand?view=net-6.0) (6) which represents an SQL statement that can be executed on a database.
```

LaPeste commented on tutorials/persistance-comparison-series/Sqlite/sqlite.md:
````diff
@@ -0,0 +1,390 @@
+"CREATE TABLE IF NOT EXISTS HitCountTableSimple (id INTEGER PRIMARY KEY, hits INTEGER )"
+```
+
+So, what does this statement mean? First we need to state what we want to do, which is `CREATE TABLE IF NOT EXISTS`, then we neede to name this table, which will just be the same as the script we are working on right now: `HitCountTableSimple`.
````
```suggestion
So, what does this statement mean? First we need to state what we want to do, which is `CREATE TABLE IF NOT EXISTS`, then we need to name this table, which will just be the same as the script we are working on right now: `HitCountTableSimple`.
```

papafe replied:
```suggestion
So, what does this statement mean? First we need to state what we want to do, which is `CREATE TABLE IF NOT EXISTS`, then we need to name this table, which will just be the same as the script we are working on right now: `HitCountTableSimple`.
```

LaPeste commented on tutorials/persistance-comparison-series/Sqlite/sqlite.md:
```diff
@@ -0,0 +1,390 @@
+
+So, what does this statement mean? First we need to state what we want to do, which is `CREATE TABLE IF NOT EXISTS`, then we neede to name this table, which will just be the same as the script we are working on right now: `HitCountTableSimple`.
+
+Last but not least we need to define how this new table is supposed to look like. This is done by naming all columns as a tuple, like `(id INTEGER PRIMARY KEY, hits INTEGER )`, with the column name (the `key`) being the first variable (`id` and `hits`), followed by the type (`INTEGER`) and any additional attributes like `PRIMARY KEY`.
```
> with the column name (the `key`) being 

I'm confused here by (the `key`). What are you referring to? If database keys, I'm a bit puzzled. Those columns to my knowledge are called `attribute`. What is that you wanted to say?

LaPeste commented on tutorials/persistance-comparison-series/Sqlite/sqlite.md:
````diff
@@ -0,0 +1,390 @@
+"INSERT OR REPLACE INTO HitCountTableSimple (id, hits) VALUES (0, " + hitCount + ")"
+```
+
+Let's decypher this one too: `INSERT OR REPLACE INTO` adds a new variable to a table or updates it, if it already exists. Next it the table name that we want to insert into, `HitCountTableSimple`. This is followed by a tuple of columns (using their key) that we would like to change, `(id, hits)`. The statement `VALUES (0, " + hitCount + ")` then defines values that should be inserted, also as a t ...
````
```suggestion
Let's decypher this one too: `INSERT OR REPLACE INTO` adds a new variable to a table or updates it, if it already exists. Next is the table name that we want to insert into, `HitCountTableSimple`. This is followed by a tuple of columns (using their key) that we would like to change, `(id, hits)`. The statement `VALUES (0, " + hitCount + ")` then defines values that should be inserted, also as a tuple. In this case we just choose `0` for the key and use whatever the current `hitCount` is as the value.
```

LaPeste replied:
> ...tuple of columns (using their key)...

> ...for the key and use whatever...

Again, I'm confused by the usage of the word "key".

papafe commented on tutorials/persistance-comparison-series/Sqlite/sqlite.md:
```diff
@@ -0,0 +1,390 @@
+
+## Introduction
+
+Our journey of exploring options given to use when it comes persistence in Unity will in this part lead to databases. More specificaclly: Sqlite.
```
Personal preference, but most of the time I've seen it spelled SQLite

DominicFrei replied:
Which is actually the correct way. I'll adjust that. 👍

LaPeste commented on tutorials/persistance-comparison-series/Sqlite/sqlite.md:
```diff
@@ -0,0 +1,390 @@
+
+> ExecuteReader is used for any result set with multiple rows/columns (e.g., SELECT col1, col2 from sometable ). ExecuteNonQuery is typically used for SQL statements without results (e.g., UPDATE, INSERT, etc.).
+
+ALl that's left to do is properly `Close()` (12) the database.
```
```suggestion
All that's left to do is to properly `Close()` (12) the database.
```

LaPeste commented on tutorials/persistance-comparison-series/Sqlite/sqlite.md:
```diff
@@ -0,0 +1,390 @@
+        dbCommandReadValues.CommandText = "SELECT * FROM HitCountTableSimple"; // 16
+        IDataReader dataReader = dbCommandReadValues.ExecuteReader(); // 17
+
+        while (dataReader.Read())
```
```suggestion
        while (dataReader.Read()) // 18
```

LaPeste commented on tutorials/persistance-comparison-series/Sqlite/sqlite.md:
````diff
@@ -0,0 +1,390 @@
+"SELECT * FROM HitCountTableSimple"
+```
+
+In this case, `SELECT` stands for `read the following values`, followed by a `*` which indicates to read all data. The keyword `FROM` than specifie the table that should be read from which is again `HitCountTableSimple`. Finally we execute this command using `ExecuteReader()` (17) since we expect data back. This data is saved in an [`IDataReader`](https://docs.microsoft.com/en-us/dotnet/api/syste ...
````
```suggestion
In this case, `SELECT` stands for `read the following values`, followed by a `*` which indicates to read all the data. The keyword `FROM` then specifies the table that should be read from which is again `HitCountTableSimple`. Finally we execute this command using `ExecuteReader()` (17) since we expect data back. This data is saved in an [`IDataReader`](https://docs.microsoft.com/en-us/dotnet/api/system.data.idatareader?view=net-6.0), from the documentation:
```

LaPeste commented on tutorials/persistance-comparison-series/Sqlite/sqlite.md:
```diff
@@ -0,0 +1,390 @@
+
+> Provides a means of reading one or more forward-only streams of result sets obtained by executing a command at a data source, and is implemented by .NET data providers that access relational databases.
+
+The way this data is read is row by row. Each time we call `dataReader.Read()` (18) we read another row from the table. Since we know there is only one row in the table we can just assign the `value` of that row to the `hitCount`. The `value` is of type `INTEGER` so we need to use [`GetInt32(1)`](https://docs.microsoft.com/en-us/dotnet/api/system.data.idatarecord.getint32?view=net-6.0) to read it ...
```
Not being that well familiar with `IDataReader`, the appearance of the word "index" threw me off. I had to go to the documentation to understand what it's meant with it.
Is there something that could be done here to briefly brush up on `IDataReader`? I'd be happy with a simple line like: `IDataReader` addresses its content in an index fashion, where the ordering matches that one of the attributes in the SQL table of interest. So in our case `id` has index 0, and `hitCount` has index 1.

Something like this would have made it a lot clearer for me. If I'm the only one, then disregard this.

papafe commented on tutorials/persistance-comparison-series/Sqlite/sqlite.md:
````diff
@@ -0,0 +1,390 @@
+Then, open a file explorer in your Unity Hub installation directory, head to the following sub directory:
+
+```
+Unity/Hub/Editor/2021.2.11f1/Editor/Data/MonoBleedingEdge/lib/mono/unity
````
A question out of ignorance. This "2021.2.11f1", is it the Unity version?

DominicFrei replied:
:D 

It indeed is, yes.

LaPeste commented on PersistenceComparison/Assets/Scripts/Sqlite/SqliteExampleSimple.cs:
```diff
@@ -0,0 +1,58 @@
+using System.Data; // 1
+using UnityEngine;
+
+public class SqliteExampleSimple : MonoBehaviour
```
I have no experience at all with SQLite and I don't have that much experience with sql databases in general either. So it's mostly curiosity: what's the reason for this not to be cached, instead of opening and closing it for every interaction? Does it have to do with persistence of data to disk?

DominicFrei replied:
`Open()` locks the db, `Close()` releases the lock. It's basically like our `CommitTransaction()`. If we don't close here when done with the click we might not close it at all and it's locked and no other part of the application can write to it.

LaPeste commented on tutorials/persistance-comparison-series/Sqlite/sqlite.md:
````diff
@@ -0,0 +1,390 @@
+}
+```
+
+The first part works basically unchanged by creating a `IDbConnection` (1) and a `IDbCommand` (2) and then read all rows again with `SELECT *` (3) but this time from `HitCountTableExtended`, finished by actually executing the command with `ExecuteReader()`
````
```suggestion
The first part works basically unchanged by creating a `IDbConnection` (1) and a `IDbCommand` (2) and then reading all rows again with `SELECT *` (3) but this time from `HitCountTableExtended`, finished by actually executing the command with `ExecuteReader()`
```

LaPeste commented on tutorials/persistance-comparison-series/Sqlite/sqlite.md:
```diff
@@ -0,0 +1,390 @@
+
+## Conclusion
+
+Sqlite is one of the options when it comes to persistence. If you have read the previous tutorials you have noticed that using it might at first seem a bit more complicated than the simple `PlayerPrefs`. You have to learn an additional 'language' to be able to communicate with your database. And due to the nature of SQL no being the easiest format to read, it might seem a bit intimmidating at fir ...
```
```suggestion
Sqlite is one of the options when it comes to persistence. If you have read the previous tutorials you have noticed that using it might at first seem a bit more complicated than the simple `PlayerPrefs`. You have to learn an additional 'language' to be able to communicate with your database. And due to the nature of SQL not being the easiest format to read, it might seem a bit intimidating at first.
```

LaPeste commented on tutorials/persistance-comparison-series/Sqlite/sqlite.md:
```diff
@@ -0,0 +1,390 @@
+
+Sqlite is one of the options when it comes to persistence. If you have read the previous tutorials you have noticed that using it might at first seem a bit more complicated than the simple `PlayerPrefs`. You have to learn an additional 'language' to be able to communicate with your database. And due to the nature of SQL no being the easiest format to read, it might seem a bit intimmidating at fir ...
+
+But the world of databases offers a lot more than can be shown in a short tutorial like this!
```
Is it necessary for this to be split in a paragraph of its own?

LaPeste commented on tutorials/persistance-comparison-series/Sqlite/sqlite.md:
```diff
@@ -0,0 +1,390 @@
+
+One of the downsides of plain files or `PlayerPrefs` that we've seen was having data in a structured way. Especially when it gets more complicated or relationships between objects should be drawn. We looked at JSON as a way to improve that situation but as soon as we need to change the format and migrate our structure it gets quite complicated. Encryption is another topic that might be important  ...
+
+In the next tutorial we will look at another database, the Realm Unity SDK, which offers similar advantages like Sqlite, while being very easy to use at the same time.
```
```suggestion
In the next tutorial we will look at another database, the Realm Unity SDK, which offers similar advantages to Sqlite, while being very easy to use at the same time.
```

LaPeste reviewed (changes requested):
Assuming that the reader does not necessarily need to be well versed in sql databases, the paragraph **Squilte** is really dense of information. And the fact that the code is all together in 1 block makes it hard to follow, as one has to keep going back and forth. In fact, the next paragraph (**Extended example**) does it differently, by showing piece by piece. This makes it much easier to follow in my opinion.
While I understand that the reader may also need to see the code all together, I still think that there's value in duplication.

Additionally, always assuming that the reader is not super familiar with sql databases, I think that the information in **Squilte** although being dense is given sparsely. This makes it hard to piece together the steps necessary for dealing with a database.
What could help in my opinion would be to have a numbered list toward the beginning of the paragraph stating the necessary steps. This prepares the reader to enter the deeper explanation of each step with an overview in mind.

About the spelling errors, I think it's most beneficial if you use a spellchecker. If what you use to write markdown is VSCode, there are a lot of plugins for that.

DominicFrei commented:
> 

That's a good point. All the other tutorials are like that too so I'm kind of inclined to live it like that for now and take that on board for the final part: Realm!

DominicFrei commented:
I have rebased this on `persistence-comparison` (which it was not) to solve the conflicts and then added two new commits, addressing all the comments.

LaPeste commented on tutorials/persistance-comparison-series/Sqlite/sqlite.md:
```diff
@@ -0,0 +1,388 @@
+
+> Provides a means of reading one or more forward-only streams of result sets obtained by executing a command at a data source, and is implemented by .NET data providers that access relational databases.
+
+`IDataReader` addresses its content in an index fashion, where the ordering matches one of the columns in the SQL table. So in our case, `id` has index 0, and `hitCount` has index 1. The way this data is read is row by row. Each time we call `dataReader.Read()` (18), we read another row from the table. Since we know there is only one row in the table, we can just assign the `value` of that row to ...
```
```suggestion
`IDataReader` addresses its content in an index fashion, where the ordering matches the one of the columns in the SQL table. So in our case, `id` has index 0, and `hitCount` has index 1. The way this data is read is row by row. Each time we call `dataReader.Read()` (18), we read another row from the table. Since we know there is only one row in the table, we can just assign the `value` of that row to the `hitCount` using its index 1. The `value` is of type `INTEGER` so we need to use [`GetInt32(1)`](https://docs.microsoft.com/en-us/dotnet/api/system.data.idatarecord.getint32?view=net-6.0) to read it and specify the index of the field we want to read as a parameter, `id` being `0` and `value` being `1`.
```

LaPeste commented on tutorials/persistance-comparison-series/Sqlite/sqlite.md:
```diff
@@ -0,0 +1,388 @@
+
+SQLite is one of the options when it comes to persistence. If you've read the previous tutorials, you've noticed that using it might at first seem a bit more complicated than the simple `PlayerPrefs`. You have to learn an additional "language" to be able to communicate with your database. And due to the nature of SQL not being the easiest format to read, it might seem a bit intimidating at first. ...
+
+One of the downsides of plain files or `PlayerPrefs` that we've seen was having data in a structured way—especially when it gets more complicated or relationships between objects should be drawn. We looked at JSON as a way to improve that situation but as soon as we need to change the format and migrate our structure, it gets quite complicated. Encryption is another topic that might be important  ...
```
```suggestion
One of the downsides of plain files or `PlayerPrefs` that we've seen was having data in a structured way—especially when it gets more complicated or relationships between objects should be drawn. We looked at JSON as a way to improve that situation but as soon as we need to change the format and migrate our structure, it gets quite complicated. Encryption is another topic that might be important for you — `PlayerPrefs` and `File` are not safe and can easily be read. Those are just some of the areas a database like SQLite might help you achieve the requirements you have for persisting your data.
```

LaPeste reviewed (changes requested):
All my feedback was addressed. There are left just 2 last very changes to be made. After that I think it's ready.
